MINIFICPP-1002 - Add SourceInitiatedSubscriptionListener processor

bakaid opened the pull request:
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

bakaid commented:
This processor has only been tested with a Windows 10 machine, and even there some weird edge cases happen where after reconnection it takes a really long time to resume proper operations, so more testing (and testing at scale) is definitely required.
Currently the processor emits flow files as Event XMLs come. I am not sure what the benefits and disadvantages of adding something like a moodycamel::ConcurrentQueue to collect these XMLs and emit them in onTrigger.

bakaid commented on thirdparty/openwsman/openwsman.patch:
```diff
@@ -0,0 +1,102 @@
+  * preset, hence marking them as weak symbols and testing to see
+  * if they are resolved before using them.
+  */
++#if 0
```
Unfortunately exporting all symbols with whole-archive methods don't work well with the weak symbol approach this library tries to take, so I had to disable these symbols entirely.

bakaid commented on centos.sh:
```diff
@@ -155,6 +155,8 @@ build_deps(){
                         INSTALLED+=("bzip2-devel")
                     elif [ "$FOUND_VALUE" = "libssh2" ]; then
                         INSTALLED+=("libssh2-devel")
+                    elif [ "$FOUND_VALUE" = "libxml2" ]; then
```
We can also bundle libxml2, but it is a pretty common library, so I left it an external dependency for the time being.

bakaid replied:
libxml2 is bundled in the new version.

bakaid commented:
@arpadboda @szaszm this is rebased, refactored and ready for review.

szaszm commented:
> @arpadboda @szaszm this is rebased, refactored and ready for review.

So much good stuff. Will review after the rocksdb state storage PR, most likely starting today.

szaszm commented on extensions/openwsman/processors/SourceInitiatedSubscriptionListener.cpp:
```diff
@@ -0,0 +1,918 @@
+
+  sendResponse(conn, machine_id, req_info->remote_addr, xml_buf, xml_buf_size);
+
+  ws_xml_free_memory(xml_buf);
+  ws_xml_destroy_doc(response);
```
I suggest using RAII for any kind of cleanup to avoid problems with exception safety, like `request`.

bakaid replied:
There is not much that can throw in here, and even those mostly on bad_alloc, but added response to a ScopeGuard.

bakaid commented on extensions/openwsman/processors/SourceInitiatedSubscriptionListener.cpp:
```diff
@@ -0,0 +1,918 @@
+}
+
+std::string SourceInitiatedSubscriptionListener::Handler::millisecondsToXsdDuration(int64_t milliseconds) {
+  char buf[1024];
+  snprintf(buf, sizeof(buf), "PT%lld.%03lldS", milliseconds / 1000, milliseconds % 1000);
+  return buf;
```
And 3 bytes for the milliseconds. That's why I chose a sensibly large buffer instead of potentially miscalculating the required size, although I admit it is more than enough.

bakaid commented on extensions/openwsman/processors/SourceInitiatedSubscriptionListener.cpp:
```diff
@@ -0,0 +1,918 @@
+}
+
+bool SourceInitiatedSubscriptionListener::loadState() {
+  std::map<std::string /*machineId*/, WsXmlDocH /*bookmark*/> bookmarks;
```
We don't need ordering, but if we had to communicate with enough machines to cause a significant performance difference between map and unordered_map, we would have bigger problems.
Doesn't hurt to change it, though, so I will.

bakaid commented on extensions/openwsman/processors/SourceInitiatedSubscriptionListener.h:
```diff
@@ -0,0 +1,173 @@
+
+  static constexpr char const* ProcessorName = "SourceInitiatedSubscriptionListener";
+
+  SourceInitiatedSubscriptionListener(std::string name, utils::Identifier uuid = utils::Identifier());
```
Done.

bakaid commented on cmake/BundledLibXml2.cmake:
```diff
@@ -0,0 +1,94 @@
+                INSTALL_COMMAND make install
+                BUILD_BYPRODUCTS "${BINARY_DIR}/thirdparty/libxml2-install/${BYPRODUCT}"
+                CONFIGURE_COMMAND ""
+                PATCH_COMMAND ./configure --enable-shared=no --enable-static=yes --with-iconv=no --with-zlib=no --with-lzma=no --with-python=no --with-ftp=no --with-http=no --prefix=${BINARY_DIR}/thirdparty/libxml2-install
```
`CONFIGURE_COMMAND` in a CMake ExternalProject would be the CMake command itself, with `./configure` projects we use the `PATCH_COMMAND` - not sure why, this is inherited practice from well before the third party refactor, but it seems to work fine. If we want to change it we would to review and test is extensively and change it everywhere - but I see no reason to do so right now.

szaszm replied:
Thanks for the explanation, I didn't know that this is common practice.

szaszm commented on CMakeLists.txt:
```diff
@@ -480,6 +480,18 @@ if ((ENABLE_ALL OR ENABLE_SFTP) AND NOT DISABLE_CURL)
+	include(BundledOpenWSMAN)
+	use_bundled_openwsman(${CMAKE_CURRENT_SOURCE_DIR} ${CMAKE_CURRENT_BINARY_DIR})
+
+    createExtension(OPENWSMAN-EXTENSIONS "OPENWSMAN EXTENSIONS" "This enables Opwensman support" "extensions/openwsman")
```
typo: s/Opwensman/Openwsman/

bakaid replied:
Thanks, fixed.

bakaid commented:
@arpadboda @szaszm 
Rebased to latest master and rewrote state storage to the new centralized one.
Also found an error in ListSFTP's and TailFile's state parsing logic (that would only manifest itself if the state became inconsistent, that's why it didn't surface), hence the changes there.
